grpc/write_job.cpp

```diff
@@ -0,0 +1,149 @@
+	std::size_t &json_offset, std::size_t &data_offset) {
+
+	auto fb_request = rpc_request.GetRoot();
+	auto fb_header = fb_request->header();
+	if (first && fb_header) {
+		deserialize_header(fb_header, request);
+	}
+	put_data_part(request.json, json_offset, fb_request->json());
+	put_data_part(request.data, data_offset, fb_request->data());
+}
+
+write_job_t::rpc_response_t serialize(const write_job_t::response_t &response) {
+	flatbuffers::grpc::MessageBuilder builder;
+
+	auto json_timestamp = to_rpc_time(response.json_timestamp);
+	auto data_timestamp = to_rpc_time(response.data_timestamp);
+
+	auto fb_response = fb_grpc_dnet::CreateLookupResponse(
+		builder,
+		serialize_cmd(builder, response.cmd),
+		response.record_flags,
+		response.user_flags,
+		builder.CreateString(response.path),
+		&json_timestamp,
+		response.json_offset,
+		response.json_size,
+		response.json_capacity,
+		builder.CreateVector(response.json_checksum),
+		&data_timestamp,
+		response.data_offset,
+		response.data_size,
+		builder.CreateVector(response.data_checksum)
+	);
+
+	builder.Finish(fb_response);
+
+	return builder.ReleaseMessage<fb_grpc_dnet::LookupResponse>();
+}
+
+} // namespace
+
+write_job_t::write_job_t(dnet_node &node, ::grpc::ServerCompletionQueue &completion_queue,
+	fb_grpc_dnet::Elliptics::AsyncService &service)
+: async_reader_(&ctx_)
+, node_(node)
+, completion_queue_(completion_queue)
+, async_service_(service) {
+	async_service_.RequestWrite(&ctx_, &async_reader_, &completion_queue_, &completion_queue_, this);
+}
+
+void write_job_t::proceed(bool more) {
+	// TODO: catch exceptions here
+	switch (state_) {
+	case state_t::REQUEST_WAITING_FIRST:
+		state_ = state_t::REQUEST_WAITING_NEXT;
+		new write_job_t(node_, completion_queue_, async_service_);
+		read_next(true /*first*/, more);
+		break;
+	case state_t::REQUEST_WAITING_NEXT:
+		read_next(false /*first*/, more);
+		break;
+	case state_t::RESPONSE_COMPLETE:
+		delete this;
+		break;
+	}
+}
+
+void write_job_t::read_next(bool first, bool more) {
+	if (first) {
+		// TODO: Use constructor or clear memory for request
+		request_ = std::make_unique<request_t >();
+	}
+	if (!more) {
+		push_request();
+		return;
+	}
+	rpc_request_t rpc_request;
+	async_reader_.Read(&rpc_request, this);
+	deserialize_part(rpc_request, first, *request_, request_json_offset_, request_data_offset_);
+}
+
+void write_job_t::push_request() {
+	if (request_json_offset_ != request_->json.size() || request_data_offset_ != request_->data.size()) {
+		// TODO: no throw
+		throw std::invalid_argument("Incomplete buffer");
+	}
+	request_->deadline = to_dnet_time(ctx_.deadline());
+
+	// TODO add to req
+	// auto responder = new responder_t(*this);
+
+	auto req = new dnet_io_req;
+	memset(req, 0, sizeof(dnet_io_req));
+	req->st = node_.st;
+	req->header = static_cast<void*>(&request_->cmd);
+	req->n2_msg = request_.release();
```

Comment:
The same as for read_job:
* is there any guarantee that `req` and req->n2_msg will be destroyed by `delete`
* `req->hsize` should be set to the size of memory pointed by `req->header`